src/serac/physics/base_physics.cpp

```diff
@@ -69,7 +67,7 @@ std::unique_ptr<FiniteElementState> BasePhysics::generateParameter(const std::st
   return new_state;
 }
 
-void BasePhysics::setParameter(size_t parameter_index, FiniteElementState& parameter_state)
+void BasePhysics::registerParameter(size_t parameter_index, const FiniteElementState& parameter_state)
```

Comment:
The problem is that I wrote this so a physics module doesn't own the parameters - a user does. This avoids a copy of a piece of field data, but it does make memory ownership patterns more irregular. I will sketch what it would look like if we just owned the parameters and set them by value. The cleaner interface might be worth the memory hit.

Reply:
Hey, I'd like to push on this so we get the discussion moving again.  Seems like there are a variety of design decisions to be make regarding a more general interface (also, support for parameters which are quadrature fields, adjoint loads which are d_qoi_d_some_rate, etc.).  Maybe we can have a meeting with a small-ish group to make some initial decisions?

Reply:
After some thought, I might actually lean toward the second solution outlined.  The reason: after advanceTimestep is called, how does the user know which states are valid to now 'get'.  Can the user get the adjoint_temperature?  Because if they do, it will be invalid.  Alternatively, if the return of advanceTimestep is a class/struct will a smaller set of getters (or member variables), then it is clear what fields are computed during advanceTimestep.  Similarly, solveAdjoint could return a struct with handles to adjoint relative fields, but probably also primal fields as well.  This again makes it clear what the computed outputs from this operator are.

Reply:
I spoke with @tupek2  yesterday afternoon about some of the ideas he suggested above, and one of the takeaways for me was that the current way we're managing the material-specific internal variables at each quadrature point is clumsy and complicating some of the interface choices. So, I'll look at coming up with a different container that can hopefully address those issues, specifically:
- no more shared ownership of quadrature data
- relaxing type safety enough to make specifying quadrature data as inputs/outputs simpler

Reply:
Also, I was going to try prototyping a slightly more 'functional' interface (in the sense of LISP, ocaml, jax, haskell, etc.) to the advanceTimestep, adjoint calls.  Ideally this will allow us to pass in the same arguments 2 times and get the same result both times!  This is also a useful first step if we plan to do dynamic checkpointing down the line, and I think will make it easier to reason about the advance logic and adjoint math.